File: EssentialChat/src/main/java/essential/chat/Commands.java
package essential.chat;

import arc.Core;
import essential.database.data.PlayerData;
import essential.util.UtilsKt;
import ksp.command.ClientCommand;
import mindustry.Vars;
import mindustry.gen.Call;
import mindustry.gen.Playerc;

/**
 * Commands implementation for the EssentialChat plugin.
 */
public class Commands {
    private static final String PLAYER_NOT_FOUND = "player.not.found";

    /**
     * Chat with a special prefix.
     *
     * @param playerData The player data
     * @param args The command arguments
     */
    @ClientCommand(name = "me", parameter = "<text...>", description = "Chat with special prefix")
    public void me(PlayerData playerData, String[] args) {
        if (playerData.getChatMuted()) {
            return;
        }

        ChatConfig conf = Main.conf;
        if (conf.getBlacklist().isEnabled()) {
            String[] blacklist = Core.settings.getDataDirectory().child("chat_blacklist.txt").readString("UTF-8").split("\r\n");
            String message = args[0];
            
            for (String s : blacklist) {
                if ((conf.getBlacklist().isRegex() && message.matches(s)) ||
                    (!conf.getBlacklist().isRegex() && message.contains(s))) {
                    playerData.err("event.chat.blacklisted");
                    return;
                }
            }
        }

        String message = args[0];
        Call.sendMessage("[orange]*[]" + Vars.netServer.chatFormatter.format(playerData.getPlayer().as(), message));
    }

    /**
     * Send a private message to another player.
     *
     * @param playerData The player data
     * @param args The command arguments
     */
    @ClientCommand(name = "pm", parameter = "<player> <message...>", description = "Send a private message")
    public void pm(PlayerData playerData, String[] args) {
        if (playerData.getChatMuted()) {
            return;
        }

        String targetName = args[0];
        Playerc target = UtilsKt.findPlayers(targetName);

        if (target == null) {
            playerData.err(PLAYER_NOT_FOUND);
        } else if (args.length > 1) {
            String message = args[1];
            playerData.getPlayer().sendMessage("[green][PM] " + target.plainName() + "[yellow] => [white] " + message);
            target.sendMessage("[blue][PM] [gray][" + playerData.getEntityId() + "][]" + playerData.getPlayer().plainName() + "[yellow] => [white] " + message);

            // This part is commented out as it requires access to database and Permission which we don't have proper references for
            /*
            database.getPlayers().stream().filter { p ->
                Permission.INSTANCE.check(p, "pm.other") && p.uuid != player.uuid() && target.uuid() != player.uuid()
            }.forEach { p ->
                p.player.sendMessage("[sky]${player.plainName()}[][yellow] => [pink]${target.plainName()} [white]: ${message}")
            }
            */
        } else {
            playerData.err("command.pm.message");
        }
    }
}

File: EssentialChat/src/main/kotlin/essential/chat/Commands.kt
package essential.chat

import arc.Core
import essential.database.data.PlayerData
import essential.util.findPlayers
import ksp.command.ClientCommand
import mindustry.Vars
import mindustry.gen.Call
import mindustry.gen.Playerc

class Commands {
    companion object {
        private const val PLAYER_NOT_FOUND = "player.not.found"
    }
    @ClientCommand(name = "me", parameter = "<text...>", description = "Chat with special prefix")
    fun me(playerData: PlayerData, arg: Array<out String>) {
        if (playerData.chatMuted) return

        val conf = Main.Companion.conf
        if (conf.blacklist.enabled) {
            val file: Array<String> = Core.settings.dataDirectory.child("chat_blacklist.txt").readString("UTF-8").split("\r\n").toTypedArray()
            for (s in file) {
                val message = arg[0]
                if ((conf.blacklist.regex && message.matches(s.toRegex())) || (!conf.blacklist.regex && message.contains(s))) {
                    playerData.err("event.chat.blacklisted")
                    return
                }
            }
        }

        val message = arg[0]
        Call.sendMessage("[orange]*[]" + Vars.netServer.chatFormatter.format(playerData.player.`as`(), message))
    }

    @ClientCommand(name = "pm", parameter = "<player> <message...>", description = "Send a private message")
    fun pm(playerData: PlayerData, arg: Array<out String>) {
        if (playerData.chatMuted) return

        val targetName = arg[0]
        val target: Playerc? = findPlayers(targetName)

        if (target == null) {
            playerData.err(PLAYER_NOT_FOUND)
        } else if (arg.size > 1) {
            val message = arg[1]
            playerData.player.sendMessage("[green][PM] " + target.plainName() + "[yellow] => [white] " + message)
            target.sendMessage("[blue][PM] [gray][" + playerData.entityId + "][]" + playerData.player.plainName() + "[yellow] => [white] " + message)

            // This part is commented out as it requires access to database and Permission which we don't have proper references for
            /*
            database.getPlayers().stream().filter { p ->
                Permission.INSTANCE.check(p, "pm.other") && p.uuid != player.uuid() && target.uuid() != player.uuid()
            }.forEach { p ->
                p.player.sendMessage("[sky]${player.plainName()}[][yellow] => [pink]${target.plainName()} [white]: ${message}")
            }
            */
        } else {
            playerData.err("command.pm.message")
        }
    }
}
